loader.php

```diff
@@ -61,6 +61,29 @@ function( $asset_path ) { return 'build/' . $asset_path; },
 	return null;
 }
 
+/**
+ * Infer a base web URL for a file system path.
+ *
+ * @param string $path Filesystem path for which to return a URL.
+ * @return string|null
+ */
+function infer_base_url( string $path ) {
+	if ( strpos( $path, get_stylesheet_directory() ) === 0 ) {
+		return get_theme_file_uri( substr( $path, strlen( get_stylesheet_directory() ) ) );
+	}
+
+	if ( strpos( $path, get_template_directory() ) === 0 ) {
+		return get_theme_file_uri( substr( $path, strlen( get_template_directory() ) ) );
+	}
+
+	// Any path not known to exist within a theme is treated as a plugin path.
+	$plugin_path = plugin_dir_path( __FILE__ );
+	if ( strpos( $path, $plugin_path ) === 0 ) {
+		return plugin_dir_url( __FILE__ ) . substr( $path, strlen( $plugin_path ) );
+	}
+
+	return '';
```

Comment:
I'd generally prefer a `WP_Error` return (and maybe even `trigger_error`) for something like this, but #yolo for now.